### PR TITLE
mtl: Fix depth stencil issues

### LINKED ISSUE #2134: mtl: Investigate failing CTS tests for dynamic_states
Currently (as of https://github.com/gfx-rs/gfx/commit/68d1539f57d1b363c72c76ca46c98e7a687ba7ad and https://github.com/gfx-rs/portability/commit/2a03ef2acd90bc0f9960531457622222f6496e24) all of the CTS `dynamic_states` tests  return `Pass` or `NotSupported` on Metal, except a few depth and stencil tests:

* dEQP-VK.dynamic_state.rs_state.depth_bias
* dEQP-VK.dynamic_state.rs_state.depth_bias_clamp
* dEQP-VK.dynamic_state.ds_state.stencil_params_basic_1
* dEQP-VK.dynamic_state.ds_state.stencil_params_advanced

Let's look into this further and understand what goes wrong in these cases.

### PR DESCRIPTION
**Pull request by grovesNL**
Fixes #2134

PR checklist:
- [X] `make` succeeds (on *nix)
- [X] `make reftests` succeeds
- [X] tested examples with the following backends: Metal

Probably easier to read these commits one at a time because they all fix separate issues:

- Always set load/store actions for attachments which aren't being cleared. We need this to preserve values in shared textures, i.e. between two separate depth and stencil attachments with the same backing texture.
- During the changes in #2129, we stopped mapping load/store actions during framebuffer creation. As a result we always used "don't care" which led to unexpected images at the beginning of render passes. For example, if image clears occurred prior to beginning the render pass, we had no guarantee that the image was still cleared at the beginning of the render pass.
- The stencil write mask updates passed the wrong argument, so they were misinterpreted as read mask updates.

### CODE REVIEW

**Comment by kvark** [comment]
> As a result we always used "don't care" which led to unexpected images at the beginning of render passes.

My assumption was that the RP descriptor in the framebuffer is not used anywhere but for `begin_render_pass`, where we are already setting all the load ops. Could you explain a bit more the use case that is broken?

**Comment by grovesNL** [comment]
During `begin_render_pass` [`clear_values` is empty](https://github.com/gfx-rs/gfx/blob/master/src/backend/metal/src/command.rs#L2147), so we don't map the ops correctly. [`clear_values` doesn't seem to be required to be the same length as attachments](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkRenderPassBeginInfo.html) if `clear` isn't used as as the load op.

Maybe there is a way to simplify this further though (by changing this `.zip)` so we don't repeat the mapping?

**Comment by kvark** [comment]
@grovesNL it says:
> The array is indexed by attachment number.

So I assumed there's got to be the same number of clear values. More precisely, there is at least as many provided as the highest attachment number that is cleared, + 1. So how about we do `clear_values.chain(repeat(_whatever))` for that `zip` iterator?

**Comment by grovesNL** [comment]
Sure, seems reasonable to me

**Comment by grovesNL** [comment]
@kvark or @gfx-rs/metallists could somebody take a look at the updated commit? We extend the iterator by `repeat(None)` now, so my earlier fix to map load/store ops during framebuffer creation isn't needed anymore.

**Comment by kvark** [comment]
Bors delegate+
Please proceed after somebody reviews this.

File: src/backend/metal/src/command.rs [hunk 1]
```diff
@@ -2142,8 +2158,12 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
         };
         let mut num_colors = 0;
 
-        for (rat, clear_value) in render_pass.attachments.iter().zip(clear_values) {
-            let value = *clear_value.borrow();
+        let clear_values_iter = clear_values
+            .into_iter()
+            .map(|c| Some(*c.borrow()))
+            .chain(iter::repeat(None));
```
Comment (kvark) [inline src/backend/metal/src/command.rs]: I don't think the `Option` really helps here. The extra elements that we add are not any worse than the ones provided but not used, there is no need to try taking extra care of them. Instead, could just go with a default clear value or something.

Comment (grovesNL) [reply src/backend/metal/src/command.rs]: Fair enough, I was trying to avoid creating extra clear values but maybe it's not a big deal.

**Pull request merged**